### ISSUE TITLE
`ddev exec` to run commands from the equivalent directory in the host

### ISSUE THREAD

**Issue by shaal** [ddev/ddev#3881]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe

Some setup (i.e. Gitpod), doesn't have `composer`, `node` installed locally on the machine. Instead, these applications are only available in the ddev container.

When I run `ddev exec yarn install` I want it to behave the same way as if I had `yarn` installed locally, meaning the directory I'm currently on, is where I really want `yarn` to run, and not in the default root directory of the website.

### Describe the solution you'd like

I would suggest changing the default behavior of `ddev exec` to always follow `current_directory` as a base of where to run the command, and add a flag for the special case when a user wants to run the command in the root directory of the project.

### Additional context

Here's the code for custom host command `exec_d` that I created in DrupalPod to achieve this functionality.
https://github.com/shaal/DrupalPod/pull/83/files#diff-60b442dab83e861ef19e31979d6cf8a119bc4b817d6712e8277e2415e03c6111

**Comment by rfay** [comment]
Maybe it should just be a flag on `ddev exec` ?

**Comment by shaal** [comment]
> Maybe it should just be a flag on `ddev exec` ?

Right.
But I believe the more common use case is to run `ddev exec` from a specific directory in the container.
I'm suggesting we change the default behavior of `ddev exec` to be aware of directory it should run from , and adding a flag to run it from root directory of project (which is how it currently running)

**Comment by rfay** [comment]
The behavior change would be pretty significant, I'm pretty hesitant to do that when people have always used `working_dir` as the execution directory for `ddev exec` and `ddev ssh`.

**Comment by shaal** [comment]
👍 
Step 1 - create the new flag functionality.

Future - figure out if default behavior should change or not.

**Comment by chrisfromredfin** [comment]
I wrote a custom command to do this:

.ddev/commands/host on  master [!?⇡]
❯ cat x
#!/bin/bash
newPath=${PWD/$DDEV_APPROOT/\/var\/www\/html}
ddev exec -d $newPath $@

...in case that's at all helpful to anyone. :)

--
Christopher J. Wells
Redfin Solutions, LLC
office: 908-4-DRUPAL
fax: 207-292-1414
***@***.***
www.redfinsolutions.com

**Comment by rfay** [comment]
Note that this capability is now built into `ddev npm` and `ddev yarn`, which may solve your problem. See the [HostWorkingDir](https://ddev.readthedocs.io/en/latest/users/extend/custom-commands/#hostworkingdir-annotation-container-commands-only) annotation.

Does that do the job for you @shaal or did you have other expectations?

**Comment by shaal** [comment]
That is great. I wanted to also have a more generic solution.

**Comment by rfay** [comment]
With that, you can now make a custom command that works the way you expect...

**Comment by shaal** [comment]
@rfay can you share an example?

In DrupalPod, I created a custom command.

I wish that the same functionality could be achieved by adding a flag to ddev's `exec` command

`.ddev/commands/host/exec_dir`
```bash
#!/bin/bash

## Description: Run programs in container, based on current host directory
## Usage: exec_d [args]
## Example: "ddev exec_d yarn"

if [ -z "$1" ]; then
  echo "No argument supplied"
  exit 1
fi

ddev exec -d /var/www/html/"$(echo "$PWD" | cut -d'/' -f4-)" "$@"
```

**Comment by rfay** [comment]
Those 2 commands... In your global .ddev...

**Comment by rfay** [comment]
https://github.com/drud/ddev/blob/3703a94a9b263a52c508a8e88be5da765787127e/pkg/ddevapp/global_dotddev_assets/commands/web/yarn#L1-L10

**Comment by GuySartorelli** [comment]
It sounds like there is now the capacity to do this, yeah? Is there any remaining work to be done on this issue?

**Comment by rfay** [comment]
It doesn't do it by default, as we didn't want to break so many people's workflow. 

Not sure what the simplest way is to work on this without breaking things. Maybe we should close it, or maybe we should add a simple shell command that does it.

**Comment by GuySartorelli** [comment]
> It doesn't do it by default, as we didn't want to break so many people's workflow.

I don't think it _should_ do it by default. I've only just started using DDEV and already that change would disrupt my workflow.
But if there's a way for people to choose to do it when they want to, I think that should be enough. I'd say people wanting to do that is more likely the outlier rather than the other way around.

**Comment by rfay** [comment]
It has already had `ddev exec -d` for 100 years. So maybe that's all we need and we should close this.

**Comment by donquixote** [comment]
> It has already had ddev exec -d for 100 years. So maybe that's all we need and we should close this.

That's not ideal, I think we still need to type out the full directory.
I would prefer something like
```sh
ddev exec --current-dir pwd  # /var/www/html
cd subdir
ddev exec --current-dir pwd  # /var/www/html/subdir
```

Currently I cannot even use `-d .`, it must be an absolute dir.

**Comment by rfay** [comment]
Lots of people use this custom command or something like it. I have it as `ddev here`, put this in ~/.ddev/commands/web/here , the `HostWorkingDir` is what makes it work:

```
#!/bin/bash

## Description: Exec anything in the same directory as on the host
## Usage: here [flags] [args]
## Example: "ddev here ls"
## ExecRaw: true
## HostWorkingDir: true

bash -c "$@"
```

I do think it would be reasonable to add this to core DDEV, and it's so easy. Not sure what to name it, and don't want to break every script in the world that uses `ddev exec` by changing its behavior.

**Comment by chrisfromredfin** [comment]
I would be in support of _either_ `ddev exec --here` (a special flavor of `-d`) or a separate command, like `ddev exec-here` - but I would definitely **not** change the default behavior of `ddev exec`!

**Comment by donquixote** [comment]
I would like something that can be applied to any command.

A parameter like `--here` can be implemented for `ddev exec` first but then we can add it to `ddev ssh` and maybe other commands like `ddev composer`, `ddev phpunit`.

But I think most of these are coming from extensions/plugins, and it is not easy to have a generic implementation to tell every command to run from a subdir. Or is it?

Could we do `ddev --here xyz`? And by doing this, behave as if that subdir is the root dir?

**Comment by rfay** [comment]
> Could we do ddev --here xyz? And by doing this, behave as if that subdir is the root dir?

I'm pretty sure that affects too many things. 

Note that `ddev yarn` and `ddev npm` and a few other things already have this by default. 

`ddev composer` uses explicit `composer_root`, which is probably better, but it also provides `-d` for a specific directory. `ddev phpunit` is not a ddev command, but is one provided by another add-on, so you would take that one up there.

**Comment by donquixote** [comment]
Alright.
Then at least `ddev <command> --here` could become a pattern that can be adapted by other commands.
Instead, if we provide a new command like `ddev exechere` or just `ddev here`, this is not that easy to adapt.

**Comment by donquixote** [comment]
`ddev here "ls -lA"` works.

**Comment by rfay** [comment]
Oh, this seems to work better: ~/.ddev/commands/web/here (`bash -c "$*"`):

```bash
#!/bin/bash

## Description: Exec anything in the same directory as on the host
## Usage: here [flags] [args]
## Example: "ddev here ls"
## ExecRaw: true
## HostWorkingDir: true

bash -c "$*"
```
